### ISSUE TITLE
no-invalid-this rule allows "this" keyword in ES6 class static methods

### ISSUE THREAD

**Issue by micnic** [eslint/eslint#6967]
When using `no-invalid-this` rule in a ES6 class static method I do not get any warnings.

``` js
class Class {
    static method() {
        return this; // should show a warning here
    }
}
```

From Docs:

> This rule judges from following conditions whether or not the function is a method:
> 
>    The function is on an object literal.
>    The function is assigned to a property.
>    The function is a method/getter/setter of ES2015 Classes. **(excepts static methods)**

From my understanding static methods are not considered a valid context for `this` keyword.

**Comment by platinumazure** [comment]
Marking as enhancement request for now since it seems to be by design.

@mysticatea Any thoughts on this?

**Comment by mysticatea** [comment]
Original no-invalid-this rule had been warning static methods.
But by #4669, that behavior was addressed as a bug.

**Comment by mysticatea** [comment]
Actually, I'd like to revive the original behavior 😀
But it's a breaking change, so I think 👍  for adding an option.

**Comment by micnic** [comment]
@mysticatea does it mean that the documentation is misleading to the old behavior? (`except static methods` fragment from the rule description)

**Comment by mysticatea** [comment]
@micnic yes.

**Comment by vitorbal** [comment]
(y) for option from me too

**Comment by micnic** [comment]
+1 for the option, but I would like to have the old behavior as the default behavior (even the docs are pointing to the old behavior), because the new behavior might be a source of errors, I ran into one of them when did this thing:

```js
class A {
    static sayHelloWorld() {
        return this.otherMethod() + ' World';
    }

    static sayHello() {
        return 'Hello';
    }
}

const sayHelloWorld = A.sayHelloWord;

class B {
    sayHelloWorld() {
        sayHelloWorld();
    }
}
```

And when I called `sayHello()` method I was expecting it to not be bound to the A class context (I forgot that I used `this` in the static methods :D).

**Comment by nzakas** [comment]
`this` is completely valid in this situation, it's a reference to the constructor function, so I believe the current default is correct (the rule is supposed to flag strict mode `this` violations). I'm not sure we should actually add an option because this isn't an invalid `this` as defined by strict mode. I'd say this would have to be a new rule though I don't think it's a good fit for core. 

@micnic in the future, please use the issue template. It really helps us a lot.

**Comment by kaicataldo** [comment]
I'm closing this issue because it looks like consensus couldn't be reached. While we wish we could accommodate all requests, we have limited resources and do need to prioritize. We've found that issues failing to reach consensus after 21 days tend to never reach consensus, and as a team have decided to close such issues. This doesn't mean the idea isn't interesting or valuable, only that it's not something the team can commit to at the moment.
